File: Simon/Tests/ApiClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using BestHTTP;
using BestHTTP.Decompression.Zlib;
using BestHTTP.Forms;
using UnityEngine;
using Random = System.Random;

namespace Tests
{
    public class LastRequest
    {
        public string EndPoint;
        public RequestData RequestData;
        public bool IsResend;
    }
    
    public class ApiClient
    {
        // service
        private EncryptionHelper _encryptionHelper = new EncryptionHelper();
        MD5 md5 = MD5.Create();
        public LastRequest LastRequest;
        public RespondData RespondData;
        
        // variables
        public long nonce = 1;
        public string Host;
        public string DeviceId;
        public string Version = "1.0";
        public string VersionKey = "any-version-name";
        public int Timeout;
        public bool isSending = false;
        
        // for testing
        public bool remoteSendInvalidPayload = false;
        public bool localSendInvalidSignBase64 = false;
        public bool localSendInvalidRequestData = false;
        public bool remoteTimeout = false;
        
        // return from server
        public User user;
        public Card[] cards;
        public Card card;
        public long receivedTimestamp;
        public long serverTimestamp;
        public string token = "";

        public ApiClient(string host, string deviceId, int timeout)
        {
            Host = host;
            DeviceId = deviceId;
            Timeout = timeout;
            var random = new Random();
            nonce = random.Next(0, 1 << 30);
        }

        public long CurrentTimestamp
        {
            get { return DateTimeOffset.UtcNow.ToUnixTimeSeconds(); }
        }
        
        public long PredictedServerTimestamp
        {
            get { return serverTimestamp + (CurrentTimestamp - receivedTimestamp);  }
        }
        
        
        public bool CanResend
        {
            get { return LastRequest != null && !isSending; }
        }

        public void Resend()
        {
            if (CanResend)
            {
                Debug.Log( "resend " + LastRequest.EndPoint );
                _request(LastRequest.EndPoint, LastRequest.RequestData, isResend : true );
            }
        }

        public void Login()
        {
            var requestData = new RequestData() {deviceId = DeviceId};
            _request("/users/login", requestData, false);
        }

        public void ListCards()
        {
            var requestData = new RequestData() {};
            _request("/cards/list", requestData);
        }

        public void CreateCard(int monsterid)
        {
            var requestData = new RequestData() {monsterId = monsterid};
            _request("/cards/create", requestData);
        }
        
        public void UpdateCard(int cardId, int exp)
        {
            var requestData = new RequestData() {cardId = cardId, exp = exp};
            _request("/cards/update", requestData);
        }
        
        public void DeleteCard(int cardId)
        {
            var requestData = new RequestData() {cardId = cardId};
            _request("/cards/delete", requestData);
        }

        public string getQuery(Dictionary<string, string> dict)
        {
            string query = string.Join("&", dict.Select(x => x.Key + "=" + Uri.EscapeDataString(x.Value)).ToArray());
            return query;
        }

        public void _request(string endPoint, RequestData requestData, bool includeSession = true, bool isResend = false )
        {
            if (isSending)
            {
                return;
            }

            //settings
            isSending = true;
            LastRequest = new LastRequest();
            LastRequest.EndPoint = endPoint;
            LastRequest.RequestData = requestData;
            
            //add request data params
            requestData.timestamp = PredictedServerTimestamp;
            requestData.version = Version;
            requestData.versionKey = VersionKey;

            if (includeSession)
            {
                requestData.session = user.session;
            }

            nonce++;
            if (!isResend)
            {
                requestData.cacheKey = nonce.ToString();
            }
            else
            {
                LastRequest.IsResend = true;
            }
            
            // ------

            string json = JsonUtility.ToJson(requestData);
            
            if (localSendInvalidRequestData)
            {
                json = "hahahaha";
            }
            Debug.Log( json );
            
            // post 
            Byte[] payload = _encryptionHelper.StringToBytes(json );
            Byte[] payloadEncrypted = _encryptionHelper.Encrypt( payload, nonce ); 
            string payloadBase64 = _encryptionHelper.BytesToBase64( payloadEncrypted );
            var postForm = new RawJsonForm();
            postForm.AddField("payloadBase64", payloadBase64);
            
            // signature
            Byte[] nonceByte = _encryptionHelper.GetNonceBytes(nonce);
            Byte[] digest =  md5.ComputeHash(payloadEncrypted);
            Byte[] signMessage = new byte[8 + digest.Length];
            nonceByte.CopyTo( signMessage, 0 );
            digest.CopyTo( signMessage, 8 );
            Byte[] signedByte = _encryptionHelper.SignMessage( signMessage );
            string signedBase64 = _encryptionHelper.BytesToBase64( signedByte );
                
            // query
            Dictionary<string, string> queryDict = new Dictionary<string, string>();
            queryDict.Add("signedBase64", signedBase64);
            queryDict.Add("token", token);
            
            // testing
            if (remoteTimeout)
            {
                queryDict.Add("remoteTimeout", "true" );
            }

            if (remoteSendInvalidPayload)
            {
                queryDict.Add("remoteSendInvalidPayload", "true");
            }
            
            if (localSendInvalidSignBase64)
            {
                queryDict["signedBase64"] = "aaaaaa";
            }
            
            var query = getQuery(queryDict);
            
            
            // send request
            var url = $"{Host}{endPoint}?{query}";
            Debug.Log( url );
            HTTPRequest request = new HTTPRequest(new Uri(url), HTTPMethods.Post, OnRespond);
            request.SetForm(postForm);
            request.Timeout = TimeSpan.FromSeconds( Timeout );
            request.Send();
        }

        private void OnRespond(HTTPRequest request, HTTPResponse response)
        {
            isSending = false;

            string dataString;

            if (response == null )
            {
                Debug.Log("timeout");
                RespondData = new RespondData();
                RespondData.errorCode = ErrorCode.Timeout;
                return;
            }
            
            if ( response.Headers.ContainsKey("content-encoding") )
            {
                // normal flow
                try
                {
                    var decrypted = _encryptionHelper.Decrypt(response.Data, nonce);
                    var decompressed = _encryptionHelper.GZipDecompress(decrypted);
                    dataString = _encryptionHelper.BytesToString(decompressed);
                }
                catch ( Exception e )
                {
                    RespondData = new RespondData();
                    RespondData.errorCode = ErrorCode.FailedToDecryptServerPayload;
                    return;
                }
            }
            else
            {
                // error flow
                dataString = response.DataAsText; 
            }
            Debug.Log( dataString );
            
            RespondData = JsonUtility.FromJson<RespondData>(dataString);
            if (RespondData.errorCode == 0 )
            {
                _OnRespondSuccess(RespondData);
            }
            else
            {
                _OnRespondError(RespondData);
            }
        }

        private void _OnRespondSuccess( RespondData respondData)
        {
            if (LastRequest.EndPoint == "/users/login")
            {
                token = respondData.token;
                _UpdateUser( respondData.user );
            }
            else
            {
                Debug.Log("Success " + LastRequest.EndPoint );
                if (respondData.body.card.id > 0 )
                {
                    _UpdateCard(respondData.body.card);
                }
                if (respondData.body.cards != null )
                {
                    _UpdateCardList( respondData.body.cards );
                }
                RespondData.isCache = LastRequest.IsResend;
            }
            serverTimestamp = RespondData.timestamp;
            receivedTimestamp = CurrentTimestamp;
        }

        private void _OnRespondError(RespondData respondData)
        {
            Debug.Log(respondData.errorCode);
            if (respondData.errorCode == ErrorCode.HasCache)
            {
                _OnRespondSuccess(respondData);
            }
        }

        private void _UpdateUser(User respondUser)
        {
            user.userId = respondUser.userId;
            user.deviceId = respondUser.deviceId;
            user.session = respondUser.session;
        }

        private void _UpdateCard(Card respondCard)
        {
            card = respondCard;
        }
        
        private void _UpdateCardList(Card[] respondCards)
        {
            cards = respondCards;
        }
    }
}

File: Martin/anti-cheat/tests/ApiClient.ts
import {IUser} from '../src/entities/IUser';
import {EncryptionService} from '../src/providers/encryption.service';
import {IRespondData} from '../src/entities/IRespondData';
import {ApiError} from '../src/errors/ApiError';
import {ErrorCode} from '../src/states/ErrorCode';
import {ICard} from '../src/entities/ICard';
import {IRequestData} from '../src/entities/IRequestData';
import axios from 'axios';
import * as crypto from 'crypto';
import * as zlib from 'zlib';
import {timeout} from "rxjs/operators";

interface ILastRequest {
    endPoint: string;
    data: IRequestData;
}

export class ApiClient {
    // services
    private _encryptionService: EncryptionService;
    public nonce: number = 0;
    public timeout: number;
    public version: string = '1.0';
    public versionKey: string = 'version-key';
    public lastRequest: ILastRequest | null;

    // for testing
    public triggerInvalidServerPayload = false;
    public triggerInvalidSignature = false;
    public triggerInvalidClientPayload = false;
    public triggerTimeout = false;

    // return from server
    public user: IUser;
    public cards: ICard[];
    public receivedTimestamp: number;
    public serverTimestamp: number;
    public token: string;

    // self-define
    public localSendInvalidRequestData: boolean = false;
    public remoteSendInvalidPayload: boolean = false;
    public remoteTimeout: boolean = false;
    public localSendInvalidSignBase64: boolean = false;

    constructor(public readonly host: string, public readonly deviceId: string, requestTimeout: number) {
        this._encryptionService = new EncryptionService();
        this.timeout = requestTimeout;
        // this.nonce = Math.random() * 256 * 256 * 256 | 0;
        this.nonce = Math.random() * (1 << 30) | 0;
    }

    public get currentTimestamp(): number {
        return +new Date() / 1000 | 0;
    }

    public get predictedServerTimestamp(): number {
        const timestamp = this.currentTimestamp;
        return this.serverTimestamp + (timestamp - this.receivedTimestamp);
    }

    public async login() {
        const data = {deviceId: this.deviceId};
        const respondData = await this._request('/users/login', data);
        this.token = respondData.token;
    }

    public async listCards() {
        const data = {};
        const respondData = await this._request('/cards/list', data);
        this.cards = respondData.body.cards;
        return this.cards;
    }

    public async createCard(monsterId: number) {
        const data = {monsterId};
        const respondData = await this._request('/cards/create', data);
        this.cards = respondData.body.cards;
        return respondData;
    }

    public async updateCard(cardId: number, exp: number) {
        const data = {cardId, exp};
        const respondData = await this._request('/cards/update', data);
        this.cards = respondData.body.cards;
        return respondData;
    }

    public async deleteCard(cardId: number) {
        const data = {cardId};
        const respondData = await this._request('/cards/delete', data);
        this.cards = respondData.body.cards;
        return respondData;
    }

    public get canResend(): boolean {
        return !!(this.lastRequest);
    }

    public async resend(): Promise<IRespondData> {
        if (this.canResend) {
            return await this._request(this.lastRequest.endPoint, this.lastRequest.data);
        }
    }

    public generateNonce() {
        const nonceBuffer = Buffer.alloc(8);
        nonceBuffer.writeUInt32LE(this.nonce, 0);
        return nonceBuffer;
    }

    private async _request(endPoint: string, data: IRequestData): Promise<IRespondData> {
        // TODO: implement
        const host = 'https://nestjs-api-test.herokuapp.com';
        const instance = axios.create({
            responseType: 'arraybuffer',
            baseURL: host,
        });
        const nonceBuffer = this.generateNonce();
        this.nonce++;

        console.log(endPoint);
        console.log(data);
        if (endPoint == "/users/login") {
            // try {
                let requestData = {
                    version: this.version,
                    versionKey: this.versionKey,
                    timestamp: this.currentTimestamp,
                    deviceId: data.deviceId ? data.deviceId : ""
                };
                let payload = this._encryptionService.encrypt(Buffer.from(JSON.stringify(requestData)), nonceBuffer);
                if (this.localSendInvalidRequestData) {
                    payload = this._encryptionService.encrypt(Buffer.from(requestData.toString()), nonceBuffer);
                }
                let payloadBase64 = payload.toString("base64");
                let md5 = crypto.createHash('md5').update(payload).digest();
                let signedBase64 = this._encryptionService.signMessage(Buffer.concat([nonceBuffer, md5])).toString("base64");
                const post = {
                    payloadBase64: payloadBase64,
                    nonce: this.nonce
                };
                const query = {
                    signedBase64: signedBase64,
                    token: "",
                    remoteTimeout: this.remoteTimeout,
                    remoteSendInvalidPayload: this.remoteSendInvalidPayload
                };
                console.log(post);console.log(query);
                const sendTime = this.currentTimestamp;
                instance.defaults.timeout = this.timeout;
                const result = await instance.post('/users/login', post, {params: query}).catch(e => {
                    this.lastRequest = {endPoint: endPoint, data: data};
                    throw new ApiError(ErrorCode.Timeout, ErrorCode.Timeout.toString(), <IRespondData>null);
                });
                const returnData = result.data;
                // console.log(returnData.toString());

                // if ((this.currentTimestamp - sendTime) * 1000 > this.timeout) {
                //     this.lastRequest = {endPoint: endPoint, data: data};
                //     throw new ApiError(ErrorCode.Timeout, ErrorCode.Timeout.toString(), <IRespondData>null);
                // }

                if (returnData.toString().includes("errorCode")) {
                    let returnError = JSON.parse(returnData.toString());
                    console.log(returnError);
                    throw new ApiError(returnError.errorCode, returnError.errorCode, <IRespondData>returnError);
                }

                let returnResult: any;
                let respondPromise = await new Promise<IRespondData>(async (resolve, reject) => {
                    await zlib.gunzip(this._encryptionService.decrypt(returnData, nonceBuffer), (error, result) => {
                        if (result == undefined || result == null) {
                            reject(ErrorCode.FailedToDecryptServerPayload);
                            return new ApiError(ErrorCode.FailedToDecryptServerPayload, ErrorCode.FailedToDecryptServerPayload.toString(), <IRespondData>null);
                        }
                        const resultJson = JSON.parse(result.toString());
                        console.log(resultJson);
                        this.user = <IUser>resultJson.user;

                        resolve(resultJson);
                        returnResult = <IRespondData>resultJson;
                    });
                }).catch(e => {
                    throw new ApiError(e, e.toString(), <IRespondData>null);
                }).then(() => {
                    return returnResult;
                });
                return respondPromise;
            // } catch (e) {
            //     console.log(e);
            //     throw e;
            // }
        } else if (endPoint.startsWith("/cards/")) {
            let requestData = {
                version: this.version,
                versionKey: this.versionKey,
                timestamp: this.currentTimestamp,
                session: this.user.session ? this.user.session : "",
                cacheKey: data.cacheKey ? data.cacheKey : "",
                cardId: data.cardId ? data.cardId : "",
                monsterId: data.monsterId ? data.monsterId : "",
                exp: data.exp ? data.exp : ""
            };
            let payload = this._encryptionService.encrypt(Buffer.from(JSON.stringify(requestData)), nonceBuffer);
            let payloadBase64 = payload.toString("base64");
            let md5 = crypto.createHash('md5').update(payload).digest();
            let signedBase64 = this._encryptionService.signMessage(Buffer.concat([nonceBuffer, md5])).toString("base64");
            if (this.localSendInvalidSignBase64) {
                signedBase64 = Buffer.concat([nonceBuffer, md5]).toString("base64");
            }
            const post = {
                payloadBase64: payloadBase64,
                nonce: this.nonce
            };
            const query = {
                signedBase64: signedBase64,
                token: this.token,
                remoteTimeout: this.remoteTimeout,
                remoteSendInvalidPayload: this.remoteSendInvalidPayload
            };
            const sendTime = this.currentTimestamp;
            instance.defaults.timeout = this.timeout;
            const result = await instance.post(endPoint, post, {params: query}).catch(e => {
                this.lastRequest = {endPoint: endPoint, data: data};
                this.lastRequest.data.cacheKey = crypto.createHash('md5').update(payload).digest("hex");
                throw new ApiError(ErrorCode.Timeout, ErrorCode.Timeout.toString(), <IRespondData>null);
            });
            const returnData = result.data;

            // if ((this.currentTimestamp - sendTime) * 1000 > this.timeout) {
            // }

            if (returnData.toString().includes("errorCode")) {
                let returnError = JSON.parse(returnData.toString());
                console.log(returnError);
                throw new ApiError(returnError.errorCode, returnError.errorCode, <IRespondData>returnError);
            }

            let returnResult: any;
            let respondPromise = await new Promise<IRespondData>(async (resolve, reject) => {
                await zlib.gunzip(this._encryptionService.decrypt(returnData, nonceBuffer), (error, result) => {
                    if (result == undefined) {
                        reject(ErrorCode.FailedToDecryptServerPayload);
                        throw new ApiError(ErrorCode.FailedToDecryptServerPayload, ErrorCode.FailedToDecryptServerPayload.toString(), <IRespondData>null);
                    }
                    const resultJson = JSON.parse(result.toString());
                    console.log(resultJson);
                    if (resultJson.timestamp < this.serverTimestamp) {
                        reject(ErrorCode.InvalidTimestamp);
                        return new ApiError(ErrorCode.InvalidTimestamp, ErrorCode.InvalidTimestamp.toString(), <IRespondData>null);
                    }

                    resolve(resultJson);
                    if (data.cacheKey) {
                        resultJson.isCache = true;
                        this.lastRequest = null;
                    }
                    returnResult = <IRespondData>resultJson;
                });
            }).catch(e => {
                throw new ApiError(e, e.toString(), <IRespondData>null);
            }).then(() => {
                return returnResult;
            });
            return respondPromise;
        }
    }
}